javatests/arcs/sdk/spec/EntitySpecTest.kt

```diff
@@ -0,0 +1,161 @@
+
+import arcs.core.common.Id
+import arcs.core.data.RawEntity
+import arcs.core.data.RawEntity.Companion.NO_REFERENCE_ID
+import arcs.core.data.util.toReferencable
+import arcs.core.entity.SchemaRegistry
+import arcs.sdk.Reference
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+private typealias Foo = EntitySpecParticle_Foo
+private typealias Bar = EntitySpecParticle_Bars
+
+/** Specification tests for entities. */
+@ExperimentalCoroutinesApi
+@RunWith(JUnit4::class)
+class EntitySpecTest {
+
+    class EntitySpecParticle : AbstractEntitySpecParticle()
+
+    private lateinit var idGenerator: Id.Generator
+
+    @get:Rule
+    val harness = EntitySpecParticleTestHarness { EntitySpecParticle() }
+
+    @Before
+    fun setUp() = runBlockingTest {
+        idGenerator = Id.Generator.newForTest("session")
+        harness.start()
+    }
+
+    @Test
+    fun createEmptyInstance() {
+        val entity = Foo()
+        assertThat(entity.bool).isFalse()
+        assertThat(entity.num).isEqualTo(0.0)
+        assertThat(entity.text).isEqualTo("")
+        assertThat(entity.ref).isNull()
+    }
+
+    @Test
+    fun createWithFieldValues() = runBlockingTest {
+        val dummyRef = createBarReference(Bar(value = "dummy"))
+        val entity = Foo(
+            bool = true,
+            num = 123.0,
+            text = "abc",
+            ref = dummyRef
+        )
+        assertThat(entity.bool).isEqualTo(true)
+        assertThat(entity.num).isEqualTo(123.0)
+        assertThat(entity.text).isEqualTo("abc")
+        assertThat(entity.ref).isEqualTo(dummyRef)
+    }
+
+    @Test
+    fun ensureIdentified() {
+        val entity = Foo()
+        assertThat(entity.entityId).isNull()
+
+        entity.ensureIdentified(idGenerator, "handle")
+        val entityId = entity.entityId
+
+        // Check that the entity ID has been set to *something*.
+        assertThat(entityId).isNotNull()
+        assertThat(entityId).isNotEmpty()
+        assertThat(entityId).isNotEqualTo(NO_REFERENCE_ID)
+        assertThat(entityId).contains("handle")
+
+        // Calling it again doesn't overwrite it.
+        entity.ensureIdentified(idGenerator, "something-else")
+        assertThat(entity.entityId).isEqualTo(entityId)
+    }
+
+    @Test
+    fun copy() = runBlockingTest {
+        val ref1 = createBarReference(Bar(value = "bar1"))
+        val ref2 = createBarReference(Bar(value = "bar2"))
+        val entity = Foo(
+            bool = true,
+            num = 123.0,
+            text = "abc",
+            ref = ref1
+        )
+        entity.identify()
+
+        val entityCopy = entity.copy(
+            bool = false,
+            num = 456.0,
+            text = "xyz",
+            ref = ref2
+        )
```

Comment:
this doesn't actually keep any of the copied fields - is that because that doesn't work yet?

Reply:
Nope, it works, I just forgot to test it. Done.